### PR TITLE
Fix hasMethod() and hasProperty() when generating traits.

### PR DESCRIPTION
**Pull request by speedytwenty**
When generating document classes, and using a trait as a mappedSuperclass, the properties and methods defined by the trait get duplicated.

The call to class_exists() in `DocumentGenerator::hasProperty()` and `DocumentGenerator::hasMethod()` fails when using a trait, causing methods to be duplicated in a trait superclass.

This PR solved the issue for me.

### CODE REVIEW

**Comment by speedytwenty** [comment]
Seems to be an issue generating hydrators using traits as mapped superclasses..

**Comment by malarzm** [comment]
> When generating document classes, and using a trait as a mappedSuperclass

Do you mean something like this?

```
/** @ODM\MappedSuperclass */
trait MyTrait {  }
```

Also for the record traits should work just fine in normal scenarios, we had #297 which was later fixed in #988

**Comment by speedytwenty** [comment]
Yes, except with generating from yaml mappings! 

I think this is different from what 297/988 addressed. That seems to detect properties and methods defined by traits that are used, not mappings that are traits themselves. But maybe I'm doing something wrong?

Eg.

```
Document/ThingTrait:
    type: mappedSuperclass
    fields:
        ...
```

```
<?php
namespace Document;
trait ThingTrait
{
```

**Comment by malarzm** [comment]
Traits can not be mapped as anything as they do not exist in runtime, only things defined in the trait are moved into class using it. So unfortunately this isn't supported :)

**Comment by speedytwenty** [comment]
I don't quite understand what you mean. The "runtime" issue seems like it would apply to annotations as well (and mappedSuperclass works with annotations?). This seems really close to working minus an issue in the hyrdator/proxy generation.

In my use case, I have a ton of fields that are re-used across multiple documents. But those documents may already be extending another mappedSuperclass, which is where mapping it as a trait becomes ideal.

I've moved back to using abstract classes, but I might reopen and pursue this as a feature request sometime!

**Comment by malarzm** [comment]
Let me expand my answer a bit:

First thing, according to [docs](http://docs.doctrine-project.org/projects/doctrine-mongodb-odm/en/latest/reference/inheritance-mapping.html#mapped-superclasses) "A mapped superclass is an abstract or concrete class that provides mapping information for its subclasses" and the fact that there can be only one parent (which could be `MappedSuperclass`) is used while [building class metadata](https://github.com/doctrine/mongodb-odm/blob/df6e1ebacb82bd361794c7c691f2186f49a9022b/lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataFactory.php#L134-L154). Because of this relationship trait can't be a `MappedSuperclass` as it's not a parent to any class.

Additionally to trait not being a parent, traits does not really exist when your script is executed, behind the scenes parser takes their contents and puts into class using the trait (eventually renaming things according to `use` declaration) - you can think of it as a "copy & paste" in the PHP itself. That's why I said that traits can not be mapped as anything (but properties defined by then can). Also supposing traits could be mapped and that mapping would be carried to a class using the trait, it would lead to dangerous mapping collisions. If one trait says that is `MappedSuperclass` and another says it's `Document` then what should be done?

Summing up, traits are not good candidates for `MappedSuperclass` (nor any other class level mapping). While I understand reasons for using traits with mapped properties we need to find another way of mapping them. I know it's pretty straightforward with annotations (I do that a lot personally in my work project) as you just map the fields in trait and that's all, I guess it's YAML (and possibly XML) that miss the way to re-use once defined mapping (however I have no idea how that could be properly implemented).

**Comment by speedytwenty** [comment]
That helps clear up my misunderstanding. Thanks for elaborating, @malarzm!
